### PR TITLE
Add TestServer support for generic WebHost

### PR DESCRIPTION
**Pull request by Tratcher**
#1583 Create new test patterns for anyone using generic host.

### CODE REVIEW

File: src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+{
+    public static class HostBuilderTestServerExtensions
+    {
+        /// <summary>
+        /// Builds and Starts the Host using a TestServer.
+        /// </summary>
+        /// <param name="hostBuilder"></param>
+        /// <param name="featureCollection"></param>
+        /// <returns></returns>
+        public static async Task<TestServer> StartTestServerAsync(this IHostBuilder hostBuilder, IFeatureCollection featureCollection = null)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs]: I think we should stop making extension methods on the builder that starts. It's been a pattern that I've hardly seen anyone use and it breaks the builder flow and the host flow. How do you dispose the host in this case? It's lost...

File: src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            await host.StartAsync();
+            return testServer;
+        }
+
+        /// <summary>
+        /// Builds and Starts the Host using a TestServer.
+        /// </summary>
+        /// <param name="hostBuilder"></param>
+        /// <returns></returns>
+        public static async Task<IHost> StartTestHostAsync(this IHostBuilder hostBuilder)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs]: Same comment as above. I don't think we should do this automatically.

**Review by davidfowl — CHANGES_REQUESTED** [review]
I think if we go this route we need to let the TestServer own the IHost instance just as it owns the IWebHost instance today. That also means making the IWebHost property work with a wrapper implementation.

File: src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs]: Agreed. I think it's better to just have TestServer.StartAsync().

File: src/Microsoft.AspNetCore.TestHost/TestServer.cs [hunk 3]
```diff
@@ -49,16 +70,26 @@ public IWebHost Host
         {
             get
             {
+                if (_hostInstance == null)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.TestHost/TestServer.cs]: nit: stay consistent with getter syntax. This one checks _hostInstance == null while other uses ??.

File: src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs (hunk 2 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.TestHost/HostBuilderTestServerExtensions.cs]: The test code is very verbose without this kind of helper. We don't want this busywork in our tests.
```
            var server = new TestServer();
            var host = new HostBuilder()
                .ConfigureWebHost(webBuilder =>
                {
                    webBuilder
                        .UseServer(server)
                        .Configure(app => { });
                })
                .Build();
            await host.StartAsync();
             var response = await server.CreateClient().GetAsync("/");
```
compared to:
```

            var host = await new HostBuilder()
                .ConfigureWebHost(webBuilder => webBuilder.Configure(app => { }))
                .StartTestHostAsync();
             var response = await host.GetTestClient().GetAsync("/");
```

**Comment by Tratcher** [comment]
Note to self:
```
            var host = await new HostBuilder()
                .ConfigureWebHost(webBuilder =>
                {
                    webBuilder
                        .UseTestServer()
                        .Configure(app => { });
                })
                .StartAsync();

             var response = await host.GetTestClient().GetAsync("/");
```

**Review by jkotalik — APPROVED** [review]

**Comment by Tratcher** [comment]
Updated with more distinct types of sugar.

**Review by davidfowl — COMMENTED** [review]
Much better!

**Pull request merged**